---
title: "FoldDelay"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

This code can be used to calculate and analyse the FoldDelay of any protein as long as a PDB structure is provided. Note that the PDB structure cannot contain any gaps. Because of this, we recommend using the AlphaFold predicted model of a protein. In addition, AlphaFold models provide a confidence measurement to assess the relative position of two residues within the predicted structure, called the Predicted Aligned Error (PAE). We used this metric to filter out interactions between residues whose relative positions concerning each other are predicted with low confidence since these interactions most probably do not occur in the actual structure, as is the case for contacts with disordered regions or some contacts between distinct domains.

First, the user needs to define some variables, including the PDB file name and an average translation rate. The PDB file name should be previously saved in the folder "Structures". Proteins from prokaryotic organisms were have on average translation rates of 20 aas/s, whereas proteins form eukaryotes have on average translation rates of 5 aas/s. As default, the AlphaFold model of the E.coli protein Peptidyl-prolyl cis-trans isomerase B is provided. 

You can run the following following chunk by pressing the green play symbol below:

```{r}
#Set the PDB file name of the protein you want to analyse (must be in the folder "Structures")
pdb_file_name <- "AF-P23869-F1-model_v4.pdb"
#Set translation rate for calculation of secondary axis
transrate <- 20
```

This chunk installs and loads all the necessary R packages.

```{r}
list.of.packages <- c("ggplot2", "bio3d", "circlize", "dplyr", "tidyr", "RColorBrewer", "stringr", "tibble","ggpubr")
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)
library(ggplot2)
library(bio3d)
library(circlize)
library(dplyr)
library(tidyr)
library(RColorBrewer)
library(stringr)
library(tibble)
library(ggpubr)
code3 <- c("ALA", "ARG", "ASN", "ASP", "CYS", "GLU", "GLN", "GLY", "HIS", 
           "ILE", "LEU", "LYS", "MET", "PHE", "PRO", "SER", "THR", "TRP", 
           "TYR", "VAL","H1S","H2S","TPO","PTR","SEP","HYP")
code1 <- c("A", "R", "N", "D", "C", "E", "Q", "G", "H", "I", "L", "K", 
           "M", "F", "P", "S", "T", "W", "Y", "V","H","H","T","Y","S","P")
```

This loads the PDB file and reads it.

```{r}
pdb_path <- "Structures/"
pdb <- read.pdb(paste0(pdb_path,pdb_file_name), maxlines = -1, multi = FALSE, rm.insert = FALSE,
                  rm.alt = TRUE, ATOM.only = FALSE, hex = FALSE, verbose = TRUE)
pdb_ca <- pdb$atom
pdb_ca <- subset(pdb_ca, elety == "CA", select = c(5,6,7,9,10,11,13))
for (k in 1:length(code3)){
  pdb_ca$resid <- gsub(code3[k],code1[k],pdb_ca$resid,ignore.case=TRUE)
}
length=max(pdb_ca$resno)
colnames(pdb_ca)[c(1,3)] <- c("Amino_acid","Index")
pdb_ca <- subset(pdb_ca, select = c(1,3))
```

The next chunk calculates all residue-residue interactions in the native structure (contact map). Two residues are considered to interact if they contained non-hydrogen atoms within 6 Å. 

```{r}
chimaerarainbow <- colorRampPalette(c("blue","cyan","green","yellow","red"))
ref.cont <- cmap(pdb, dcut=6, scut=1,verbose=FALSE)
data = as.data.frame(ref.cont)
names(data) <- seq(1,length)
data$from = seq(1,length)
simple <- gather(data,"to","value",-from)%>%
    mutate(color = rep(chimaerarainbow(length),length))%>% 
    na.omit() %>%
    filter(value !=0) %>%
    select("to","from","color")
colnames(simple)[2] <- "Index"
simple <- merge(simple, pdb_ca)
colnames(simple) <- c("index_1","Index","color","aa_1")
simple <- merge(simple, pdb_ca)
colnames(simple) <- c("index_2","index_1","color","aa_1","aa_2")
simple <- subset(simple, select = c(2,4,1,5,3))
simple$index_1 <- as.numeric(simple$index_1)
simple$index_2 <- as.numeric(simple$index_2)
simple <- simple[order(simple$index_1,simple$index_2),]
simple$distance <- simple$index_2 - simple$index_1
simple$protein = "protein"
distancecol<-colorRampPalette(c('#fef0d9','#fdcc8a','#fc8d59','#e34a33','#b30000'))
simple <- simple %>%
  mutate(distancecolor = distancecol(max(simple$distance))[distance]) %>%
  arrange(index_1)
```

The next chunk of code calculates the FoldDelay of a protein. In other words, it calculates the number of residues that need to be synthesized before residue (i) can engage with all its native interaction partners. FD can also be expressed in time units by factoring the decoding times (tdec) of the different residues between i and j (here we are using an average decoding time for all residues).

The output of this code is a contact map plot (contact_map_nopae) showing the furthest away C-terminal interactor of every residue and the per-residue FD profile of our protein (folddelay_profile_nopae and folddelay_table_nopae).

Note that this chunk does not filter out any interactions (Filtering based on PAE is not applied).

```{r}
output_file <- strsplit(pdb_file_name, "[.]")[[1]][1]
dir.create(file.path(paste0("Output/",output_file)), showWarnings = FALSE)
#Calculates FD
simple_maxint <- simple %>%
  group_by(index_1) %>%
  filter(index_2==max(index_2,na.rm =T)) %>%
  ungroup() %>%
  complete(index_1=seq(1,length)) %>%
  mutate(color = chimaerarainbow(length),
         protein ="protein") %>%
  arrange(distance)
#Contact map plot
pdf(paste0("Output/",output_file,"/contact_map_nopae.pdf"), width=7.2, height=4.5)
a<-{circos.clear()
  circos.par(start.degree = 90,xaxis.clock.wise=F)
  circos.initialize(simple_maxint$protein,xlim=c(1,length))
  circos.track(ylim=c(0,1),bg.border = NA)
  circos.axis(h=0,major.at = 0)
  circos.points(y=0,x=simple_maxint$index_1,col=simple_maxint$color,pch = 19)
  for(i in 1:nrow(simple_maxint)){
    if(!is.na(simple_maxint$index_2[i])){
      circos.link(sector.index1 = "protein",point1=simple_maxint$index_1[i],sector.index2 = "protein",
                  point2 = simple_maxint$index_2[i],col=simple_maxint$distancecolor[i])
    }
  }
}
print(a)
dev.off()
#Plot for FD per-residue profile
simple_maxint <- simple_maxint[order(simple_maxint$index_1),]
simple_maxint$distance[length] <- 0
simple_maxint$aa_1[length] <- pdb_ca$Amino_acid[length]
pdf(paste0("Output/",output_file,"/folddelay_profile_nopae.pdf"), width=7.2, height=4.5)
FD_profile<-ggplot(simple_maxint, aes(x=index_1,y=distance,color=index_1)) + geom_line() + scale_color_gradientn(colors=c("blue", "cyan", "green", "yellow","red")) +
  theme_pubr() +
  theme(legend.position="none") + 
  xlab("Position in sequence") + 
  scale_y_continuous("FoldDelay (aa)",
                     sec.axis = sec_axis(~./transrate,name="FoldDelay (s)"))
print(FD_profile)
dev.off()
#Table for FD per-residue profile
fold_delay_nopae <- subset(simple_maxint, select = c(1:4,6))
fold_delay_nopae$distance2 <- fold_delay_nopae$distance / transrate
colnames(fold_delay_nopae) <- c("Position","Amino acid","Max interactor position","Max interactor amino acid","FoldDelay (aa)","FoldDelay (s)")
write.table(fold_delay_nopae, file = paste0("Output/",output_file,"/folddelay_table_nopae.tab"), quote = FALSE, sep = "\t",
            row.names = FALSE, col.names = TRUE, qmethod = c("escape", "double"))
```

This chunk, as the one above, calculates the FoldDelay of a protein. However, here interactions are filtered using the Predicted Aligned Error (PAE) metric from AlphaFold. We used this metric to filter out interactions between residues whose relative positions concerning each other are predicted with low confidence since these interactions most probably do not occur in the actual structure. Therefore, this code only works with AlphaFold models.

The output of this code is a contact map plot (contact_map_pae) showing the furthest away C-terminal interactor of every residue and the per-residue FD profile of our protein (folddelay_profile_pae and folddelay_table_pae). 

Note that for the example protein provided, both the FD profiles with and without PAE filtering are the same. This is because the structure of this single domain protein is predicted with high confidence.

```{r}
output_file <- strsplit(pdb_file_name, "[.]")[[1]][1]
dir.create(file.path(paste0("Output/",output_file)), showWarnings = FALSE)
#Download PAE file and make a matrix
uniprot_id_current <- strsplit(output_file, "-")[[1]][2]
url <- paste0("https://alphafold.ebi.ac.uk/files/AF-",uniprot_id_current,"-F1-predicted_aligned_error_v4.json")
download.file(url, paste("tmp/", "pae_file.json", sep = ""), mode = "wb")
pae_file <- read.table("~/Documents/PhD/Projects/FoldDelay/FoldDelay_2024/tmp/pae_file.json", quote="\"", comment.char="")
pae_file <- pae_file$V1
json_data <- fromJSON(pae_file)
json_matrix <- json_data$predicted_aligned_error[[1]]
#Filter out interactions with an expected position error (based on PAE) > 6 Å.
simple$flag <- "No" 
for(i in 1:nrow(simple)){
  temp_from <- simple$index_1[i]
  temp_to <- simple$index_2[i]
  temp_pae <- json_matrix[temp_from,temp_to]
  if (temp_pae > 6) {
    simple$flag[i] <- "Yes" 
  }
}
simple_pae <- subset(simple, flag == "No")
#Calculates FD with PAE
simple_maxint <- simple_pae %>%
  group_by(index_1) %>%
  filter(index_2==max(index_2,na.rm =T)) %>%
  ungroup() %>%
  complete(index_1=seq(1,length)) %>%
  mutate(color = chimaerarainbow(length),
         protein ="protein") %>%
  arrange(distance)
#Contact map plot
pdf(paste0("Output/",output_file,"/contact_map_pae.pdf"), width=7.2, height=4.5)
a<-{circos.clear()
  circos.par(start.degree = 90,xaxis.clock.wise=F)
  circos.initialize(simple_maxint$protein,xlim=c(1,length))
  circos.track(ylim=c(0,1),bg.border = NA)
  circos.axis(h=0,major.at = 0)
  circos.points(y=0,x=simple_maxint$index_1,col=simple_maxint$color,pch = 19)
  for(i in 1:nrow(simple_maxint)){
    if(!is.na(simple_maxint$index_2[i])){
      circos.link(sector.index1 = "protein",point1=simple_maxint$index_1[i],sector.index2 = "protein",
                  point2 = simple_maxint$index_2[i],col=simple_maxint$distancecolor[i])
    }
  }
}
print(a)
dev.off()
#Plot for FD per-residue profile
simple_maxint <- simple_maxint[order(simple_maxint$index_1),]
simple_maxint$distance[length] <- 0
simple_maxint$aa_1[length] <- pdb_ca$Amino_acid[length]
pdf(paste0("Output/",output_file,"/folddelay_profile_pae.pdf"), width=7.2, height=4.5)
FD_profile<-ggplot(simple_maxint, aes(x=index_1,y=distance,color=index_1)) + geom_line() + scale_color_gradientn(colors=c("blue", "cyan", "green", "yellow","red")) +
  theme_pubr() +
  theme(legend.position="none") + 
  xlab("Position in sequence") + 
  scale_y_continuous("FoldDelay (aa)",
                     sec.axis = sec_axis(~./transrate,name="FoldDelay (s)"))
print(FD_profile)
dev.off()
#Table for FD per-residue profile
fold_delay_nopae <- subset(simple_maxint, select = c(1:4,6))
fold_delay_nopae$distance2 <- fold_delay_nopae$distance / transrate
colnames(fold_delay_nopae) <- c("Position","Amino acid","Max interactor position","Max interactor amino acid","FoldDelay (aa)","FoldDelay (s)")
write.table(fold_delay_nopae, file = paste0("Output/",output_file,"/folddelay_table_pae.tab"), quote = FALSE, sep = "\t",
            row.names = FALSE, col.names = TRUE, qmethod = c("escape", "double"))
```

